File: mobile-app/lib/src/booking_api.dart
import 'dart:convert';
import 'dart:developer' as developer;
import 'package:http/http.dart' as http;
import 'booking_api_base.dart';
import 'auth_service.dart';

class BookingApi {
  Future<Map<String, String>> _authHeaders() async {
    final token = await AuthService.instance.getAccessToken();
    return {
      'Authorization': 'Bearer $token',
      'Content-Type': 'application/json',
    };
  }

  // Room endpoints
  Future<List<Map<String, dynamic>>> getRooms({
    int? capacity,
    String? location,
    List<String>? amenities,
  }) async {
    final headers = await _authHeaders();
    final queryParams = <String, String>{};
    if (capacity != null) queryParams['capacity'] = capacity.toString();
    if (location != null && location.isNotEmpty) queryParams['location'] = location;
    if (amenities != null && amenities.isNotEmpty) {
      queryParams['amenities'] = amenities.join(',');
    }

    final uri = Uri.parse('${BookingApiBase.base}/api/v1/rooms').replace(queryParameters: queryParams);
    final response = await http.get(uri, headers: headers);

    if (response.statusCode != 200) {
      throw Exception('GET /rooms failed: ${response.statusCode} ${response.body}');
    }

    final data = json.decode(response.body) as Map<String, dynamic>;
    final rooms = (data['rooms'] as List? ?? []) as List<dynamic>;
    return rooms.map((r) => r as Map<String, dynamic>).toList();
  }

  Future<Map<String, dynamic>> getRoom(int id) async {
    final headers = await _authHeaders();
    final response = await http.get(
      Uri.parse('${BookingApiBase.base}/api/v1/rooms/$id'),
      headers: headers,
    );

    if (response.statusCode != 200) {
      throw Exception('GET /rooms/$id failed: ${response.statusCode} ${response.body}');
    }

    final data = json.decode(response.body) as Map<String, dynamic>;
    return data['room'] as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getRoomAvailability(
    int roomId,
    DateTime start,
    DateTime end,
  ) async {
    final headers = await _authHeaders();
    final queryParams = {
      'start': start.toUtc().toIso8601String(),
      'end': end.toUtc().toIso8601String(),
    };

    final uri = Uri.parse('${BookingApiBase.base}/api/v1/rooms/$roomId/availability')
        .replace(queryParameters: queryParams);
    final response = await http.get(uri, headers: headers);

    if (response.statusCode != 200) {
      throw Exception('GET /rooms/$roomId/availability failed: ${response.statusCode} ${response.body}');
    }

    return json.decode(response.body) as Map<String, dynamic>;
  }

  // Booking endpoints
  Future<Map<String, dynamic>> createBooking({
    required int roomId,
    required DateTime startTs,
    required DateTime endTs,
    String? title,
    List<String>? attendees,
    String? idempotencyKey,
  }) async {
    final headers = await _authHeaders();
    if (idempotencyKey != null && idempotencyKey.isNotEmpty) {
      headers['Idempotency-Key'] = idempotencyKey;
    }

    final body = json.encode({
      'room_id': roomId,
      'start_ts': startTs.toUtc().toIso8601String(),
      'end_ts': endTs.toUtc().toIso8601String(),
      if (title != null && title.isNotEmpty) 'title': title,
      if (attendees != null && attendees.isNotEmpty) 'attendees': attendees,
    });

    final response = await http.post(
      Uri.parse('${BookingApiBase.base}/api/v1/bookings'),
      headers: headers,
      body: body,
    );

    if (response.statusCode != 200 && response.statusCode != 201) {
      final errorBody = json.decode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['message'] ?? 'Failed to create booking: ${response.statusCode}');
    }

    final data = json.decode(response.body) as Map<String, dynamic>;
    return data['booking'] as Map<String, dynamic>;
  }

  Future<List<Map<String, dynamic>>> getBookings({
    int? roomId,
    DateTime? startDate,
    DateTime? endDate,
    String? status,
  }) async {
    final headers = await _authHeaders();
    final queryParams = <String, String>{};
    if (roomId != null) queryParams['room_id'] = roomId.toString();
    if (startDate != null) {
      queryParams['start_date'] = startDate.toUtc().toIso8601String();
    }
    if (endDate != null) {
      queryParams['end_date'] = endDate.toUtc().toIso8601String();
    }
    if (status != null && status.isNotEmpty) queryParams['status'] = status;

    final uri = Uri.parse('${BookingApiBase.base}/api/v1/bookings').replace(queryParameters: queryParams);
    final response = await http.get(uri, headers: headers);

    if (response.statusCode != 200) {
      throw Exception('GET /bookings failed: ${response.statusCode} ${response.body}');
    }

    final data = json.decode(response.body) as Map<String, dynamic>;
    final bookings = (data['bookings'] as List? ?? []) as List<dynamic>;
    return bookings.map((b) => b as Map<String, dynamic>).toList();
  }

  Future<Map<String, dynamic>> getBooking(int id) async {
    final headers = await _authHeaders();
    final response = await http.get(
      Uri.parse('${BookingApiBase.base}/api/v1/bookings/$id'),
      headers: headers,
    );

    if (response.statusCode != 200) {
      throw Exception('GET /bookings/$id failed: ${response.statusCode} ${response.body}');
    }

    final data = json.decode(response.body) as Map<String, dynamic>;
    return data['booking'] as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> cancelBooking(int id) async {
    final headers = await _authHeaders();
    final response = await http.delete(
      Uri.parse('${BookingApiBase.base}/api/v1/bookings/$id'),
      headers: headers,
    );

    if (response.statusCode != 200) {
      final errorBody = json.decode(response.body) as Map<String, dynamic>;
      throw Exception(errorBody['message'] ?? 'Failed to cancel booking: ${response.statusCode}');
    }

    final data = json.decode(response.body) as Map<String, dynamic>;
    return data['booking'] as Map<String, dynamic>;
  }
}

File: admin-web/src/lib/bookingApi.ts
import axios from 'axios'
import { config } from '../config/env'

// Phase 3 API client for booking endpoints
const bookingApi = axios.create({
  baseURL: config.BOOKING_API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
})

// Add token interceptor
bookingApi.interceptors.request.use((config) => {
  const token = localStorage.getItem('access_token')
  if (token) {
    config.headers.Authorization = `Bearer ${token}`
  }
  return config
})

// Handle errors
bookingApi.interceptors.response.use(
  (response) => response,
  (error) => {
    console.error('Booking API Error:', {
      url: error.config?.url,
      method: error.config?.method,
      status: error.response?.status,
      data: error.response?.data,
      message: error.message,
    })
    
    if (error.response?.status === 401) {
      localStorage.removeItem('access_token')
      localStorage.removeItem('id_token')
      if (window.location.pathname !== '/login') {
        window.location.href = '/login'
      }
    }
    return Promise.reject(error)
  }
)

export interface Room {
  id: number
  name: string
  capacity: number
  amenities: string[]
  location: string | null
  active: boolean
  ownerTeamId: number | null
  externalCalendarId: string | null
  createdAt: string
  updatedAt: string
}

export interface Booking {
  id: number
  roomId: number
  userId: number
  startTs: string
  endTs: string
  status: 'PENDING' | 'CONFIRMED' | 'CANCELLED'
  title: string | null
  attendees: string[]
  idempotencyKey: string | null
  externalEventId: string | null
  createdAt: string
  updatedAt: string
}

export interface BlackoutWindow {
  id: number
  roomId: number
  startTs: string
  endTs: string
  reason: string | null
  createdBy: number | null
  createdAt: string
}

export interface CreateBookingInput {
  room_id: number
  start_ts: string
  end_ts: string
  title?: string | null
  attendees?: string[]
}

export interface CreateBlackoutInput {
  room_id: number
  start_ts: string
  end_ts: string
  reason?: string | null
}

export interface AvailabilityResponse {
  roomId: number
  start: string
  end: string
  bookings: Array<{
    id: number
    start: string
    end: string
    title: string | null
  }>
  blackouts: Array<{
    id: number
    start: string
    end: string
    reason: string | null
  }>
}

// Room endpoints
export async function getRooms(params?: {
  date?: string
  capacity?: number
  amenities?: string[]
  active?: boolean
  location?: string
}): Promise<{ rooms: Room[] }> {
  const queryParams: Record<string, string> = {}
  if (params?.date) queryParams.date = params.date
  if (params?.capacity) queryParams.capacity = String(params.capacity)
  if (params?.amenities?.length) queryParams.amenities = params.amenities.join(',')
  if (params?.active !== undefined) queryParams.active = String(params.active)
  if (params?.location) queryParams.location = params.location

  const response = await bookingApi.get('/api/v1/rooms', { params: queryParams })
  return response.data
}

export async function getRoom(id: number): Promise<{ room: Room }> {
  const response = await bookingApi.get(`/api/v1/rooms/${id}`)
  return response.data
}

export async function createRoom(data: {
  name: string
  capacity: number
  amenities?: string[]
  location?: string | null
  active?: boolean
  owner_team_id?: number | null
}): Promise<{ room: Room }> {
  const response = await bookingApi.post('/api/v1/rooms', data)
  return response.data
}

export async function updateRoom(
  id: number,
  data: Partial<{
    name: string
    capacity: number
    amenities: string[]
    location: string | null
    active: boolean
    owner_team_id: number | null
  }>
): Promise<{ room: Room }> {
  const response = await bookingApi.patch(`/api/v1/rooms/${id}`, data)
  return response.data
}

export async function deleteRoom(id: number): Promise<{ room: Room; message: string }> {
  const response = await bookingApi.delete(`/api/v1/rooms/${id}`)
  return response.data
}

export async function getRoomAvailability(
  id: number,
  start: string,
  end: string
): Promise<AvailabilityResponse> {
  const response = await bookingApi.get(`/api/v1/rooms/${id}/availability`, {
    params: { start, end },
  })
  return response.data
}

// Booking endpoints
export async function createBooking(
  data: CreateBookingInput,
  idempotencyKey?: string
): Promise<{ booking: Booking }> {
  const headers = idempotencyKey ? { 'Idempotency-Key': idempotencyKey } : {}
  const response = await bookingApi.post('/api/v1/bookings', data, { headers })
  return response.data
}

export async function getBooking(id: number): Promise<{ booking: Booking }> {
  const response = await bookingApi.get(`/api/v1/bookings/${id}`)
  return response.data
}

const normalizeDateParam = (value?: string, endOfDay = false): string | undefined => {
  if (!value) {
    return undefined
  }

  const hasTime = value.includes('T')

  if (!hasTime) {
    return `${value}T${endOfDay ? '23:59:59' : '00:00:00'}Z`
  }

  const date = new Date(value)
  return Number.isNaN(date.getTime()) ? value : date.toISOString()
}

export async function listBookings(params?: {
  user_id?: number
  room_id?: number
  start_date?: string
  end_date?: string
  status?: 'PENDING' | 'CONFIRMED' | 'CANCELLED'
}): Promise<{ bookings: Booking[] }> {
  const queryParams: Record<string, string> = {}
  if (params?.user_id) queryParams.user_id = String(params.user_id)
  if (params?.room_id) queryParams.room_id = String(params.room_id)
  const normalizedStart = normalizeDateParam(params?.start_date)
  const normalizedEnd = normalizeDateParam(params?.end_date, true)
  if (normalizedStart) queryParams.start_date = normalizedStart
  if (normalizedEnd) queryParams.end_date = normalizedEnd
  if (params?.status) queryParams.status = params.status

  const response = await bookingApi.get('/api/v1/bookings', { params: queryParams })
  return response.data
}

export async function cancelBooking(id: number): Promise<{ booking: Booking }> {
  const response = await bookingApi.delete(`/api/v1/bookings/${id}`)
  return response.data
}

// Blackout endpoints (ADMIN only)
export async function createBlackout(data: CreateBlackoutInput): Promise<{ blackout: BlackoutWindow }> {
  const response = await bookingApi.post('/api/v1/blackouts', data)
  return response.data
}

export async function listBlackouts(roomId?: number): Promise<{ blackouts: BlackoutWindow[] }> {
  const params = roomId ? { room_id: String(roomId) } : {}
  const response = await bookingApi.get('/api/v1/blackouts', { params })
  return response.data
}

export async function updateBlackout(
  id: number,
  data: Partial<{ start_ts: string; end_ts: string; reason: string | null }>
): Promise<{ blackout: BlackoutWindow }> {
  const response = await bookingApi.patch(`/api/v1/blackouts/${id}`, data)
  return response.data
}

export async function deleteBlackout(id: number): Promise<void> {
  await bookingApi.delete(`/api/v1/blackouts/${id}`)
}

// ICS Export
export async function exportBookingsICS(params: {
  room_id?: number
  start: string
  end: string
}): Promise<Blob> {
  const queryParams: Record<string, string> = {
    start: params.start,
    end: params.end,
  }
  if (params.room_id) queryParams.room_id = String(params.room_id)

  const response = await bookingApi.get('/api/v1/exports/bookings.ics', {
    params: queryParams,
    responseType: 'blob',
  })
  return response.data
}

export default bookingApi

